File: Leet0100_SameTree/soln.py
from typing import Optional
from typing import List

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:

        if p is None and q is None:
            return True

        return self.traverseInorder(p, q)

    def traverseInorder(self, root_p: Optional[TreeNode], root_q: Optional[TreeNode],) -> bool:

        if root_p and root_p.val == root_q and root_q.val:
            return True

        return self.traverseInorder(root_p.left, root_q.left) and self.traverseInorder(root_p.right, root_q.right)


solution = Solution()

# Define some formatting constants
TRUNC = 1000
NULL = -999
PASS_FORMAT = "\033[92m{}\033[0m"  # Green text
FAIL_FORMAT = "\033[91m{}\033[0m"  # Red text

def create_binary_tree(values: List[int]) -> Optional[TreeNode]:
    """Convert a list of integers to a binary tree.

    The input list represents a level-order traversal of the tree,
    where NULL (-999) indicates a missing node.
    """
    if not values:
        return None

    # Create the root node
    root = TreeNode(values[0])

    # Use a queue to keep track of nodes that need children
    queue = [root]
    i = 1  # Start from the second element

    # Process the rest of the values
    while queue and i < len(values):
        # Get the next node that needs children
        current = queue.pop(0)

        # Assign left child if value is not NULL
        if i < len(values) and values[i] != NULL:
            current.left = TreeNode(values[i])
            queue.append(current.left)

        i += 1

        # Assign right child if value is not NULL
        if i < len(values) and values[i] != NULL:
            current.right = TreeNode(values[i])
            queue.append(current.right)

        i += 1

    return root

def truncate_str(s: str, max_length: int = TRUNC) -> str:
    """Truncate a string to max_length characters, adding '...' if truncated."""
    if len(s) <= max_length:
        return s
    return s[:max_length-3] + "..."

def assert_solution(input_list1: List[int], input_list2: List[int], expected: bool):
    """Test solution with given input and expected output."""
    input1_copy = list(input_list1)
    input2_copy = list(input_list2)
    output = solution.isSameTree(create_binary_tree(input1_copy), create_binary_tree(input2_copy))
    assert_pass = output == expected

    result = "PASS" if assert_pass else "FAIL"
    formatted_result = PASS_FORMAT.format(result) if assert_pass else FAIL_FORMAT.format(result)

    print(
        f"{formatted_result}: "
        f"Input 1: {truncate_str(str(input_list1))}, "
        f"Input 2: {truncate_str(str(input_list2))}, "
        f"Expected: {truncate_str(str(expected))}"
        f"{' , Got: ' + truncate_str(str(output)) if not assert_pass else ''}"
    )

    return assert_pass

# Example 1 from problem statement
assert_solution([1, NULL, 2, 3], [1, 3, 2], False)

# Example 2 from problem statement
assert_solution([1, 2, 3, 4, 5, NULL, 8, NULL, NULL, 6, 7, 9], [4, 2, 6, 5, 7, 1, 3, 9, 8], False)

# Example 3: Empty tree
assert_solution([], [], True)

# Example 4: Single node tree
assert_solution([1], [1], True)

# Left-skewed tree
assert_solution([3, 2, NULL, 1], [1, 2, 3], False)

# Right-skewed tree
assert_solution([1, NULL, 2, NULL, 3], [1, 2, 3], False)

# Perfect binary tree
assert_solution([4, 2, 6, 1, 3, 5, 7], [1, 2, 3, 4, 5, 6, 7], False)

# Tree with negative values
assert_solution([0, -3, 9, -10, NULL, 5], [-10, -3, 0, 5, 9], False)

# Tree with duplicate values
assert_solution([5, 3, 8, 3, 4, 7, 8], [3, 3, 4, 5, 7, 8, 8], False)

# Tree with boundary values
assert_solution([0, -100, 100], [-100, 0, 100], False)

# Complex tree structure
assert_solution([8, 3, 10, 1, 6, NULL, 14, NULL, NULL, 4, 7, 13], [1, 3, 4, 6, 7, 8, 10, 13, 14], False)

# Deep tree - testing nested structure
assert_solution([5, 4, 8, 11, NULL, 13, 4, 7, 2, NULL, NULL, NULL, 1], [7, 11, 2, 4, 5, 13, 8, 4, 1], False)

# Unbalanced tree with multiple levels
assert_solution([10, 5, 15, 3, 7, NULL, 20, NULL, NULL, 6, 8], [3, 5, 6, 7, 8, 10, 15, 20], False)

# Base cases
assert_solution([], [], True)  # Empty trees are identical
assert_solution([1], [1], True)  # Single node trees with same value
assert_solution([1], [2], False)  # Single node trees with different values
assert_solution([1], [], False)  # One empty tree, one non-empty tree

# Simple cases - different structures
assert_solution([1, 2], [1, NULL, 2], False)  # Left child vs right child
assert_solution([1, 2, 3], [1, 3, 2], False)  # Same values but different positions
assert_solution([1, 2, NULL], [1, NULL, 2], False)  # Left child vs right child

# Simple cases - same structure but different values
assert_solution([1, 2, 3], [1, 2, 4], False)  # Difference in right child
assert_solution([1, 2, 3], [1, 4, 3], False)  # Difference in left child
assert_solution([5, 2, 3], [1, 2, 3], False)  # Difference in root

# Same trees with various structures
assert_solution([1, 2, 3], [1, 2, 3], True)  # Complete binary tree of depth 1
assert_solution([1, 2, 3, 4, 5], [1, 2, 3, 4, 5], True)  # Complete left side
assert_solution([1, NULL, 2, NULL, 3], [1, NULL, 2, NULL, 3], True)  # Right-skewed tree

# Complex cases - deep trees
assert_solution([1, 2, 3, 4, 5, 6, 7], [1, 2, 3, 4, 5, 6, 7], True)  # Perfect binary tree of depth 2
assert_solution([1, 2, 3, 4, 5, 6, 7], [1, 2, 3, 4, 5, 7, 6], False)  # One node swapped

# Edge cases using constraint boundaries
assert_solution([-10000], [-10000], True)  # Minimum value allowed
assert_solution([10000], [10000], True)  # Maximum value allowed
assert_solution([-10000], [10000], False)  # Min vs max values

# Large trees (approaching the constraint of 100 nodes)
# Create a balanced tree with many nodes but still readable
assert_solution([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
               [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15], True)  # All nodes match

# Different at leaf level
assert_solution([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
               [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 16], False)  # Last node different

# One tree deeper than the other
assert_solution([1, 2, 3, 4, 5, 6, 7], [1, 2, 3, 4, 5, 6, 7, 8], False)  # Second tree has extra node

# Mixed NULL values in different positions
assert_solution([1, 2, 3, NULL, 4, NULL, 5], [1, 2, 3, NULL, 4, NULL, 5], True)  # Same structure with NULLs
assert_solution([1, 2, 3, NULL, 4, NULL, 5], [1, 2, 3, 4, NULL, 5, NULL], False)  # Different structure with NULLs

# LeetCode fail
assert_solution([2,2,2,NULL,2,NULL,NULL,2], [2,2,2,2,NULL,2,NULL], False)

File: Leet0111_MinimumDepthofBinaryTree/soln.java
package Leet0111_MinimumDepthofBinaryTree;

import java.util.Arrays;

public class soln {

    class Solution {

        public int minDepth(TreeNode root) {

            if (root == null) {
                return 0;
            }

            return 1 + Math.min(
                    minDepth(root.left),
                    minDepth(root.right));
        }
    }

    private static final int MAX_STRING_LENGTH = 50;
    private static final String ANSI_RESET = "\u001B[0m";
    private static final String ANSI_GREEN = "\u001B[92m";
    private static final String ANSI_RED = "\u001B[91m";

    private static String greenText(String text) {
        return ANSI_GREEN + text + ANSI_RESET;
    }

    private static String redText(String text) {
        return ANSI_RED + text + ANSI_RESET;
    }

    private static TreeNode createBinaryTree(Integer[] values) {
        if (values == null || values.length == 0 || values[0] == null) {
            return null;
        }

        // Create the root node
        TreeNode root = new TreeNode(values[0]);

        // Use a queue to keep track of nodes that need children
        java.util.Queue<TreeNode> queue = new java.util.LinkedList<>();
        queue.add(root);
        int i = 1; // Start from the second element

        // Process the rest of the values
        while (!queue.isEmpty() && i < values.length) {
            // Get the next node that needs children
            TreeNode current = queue.poll();

            // Assign left child if value is not null
            if (i < values.length) {
                if (values[i] != null) {
                    current.left = new TreeNode(values[i]);
                    queue.add(current.left);
                }
                i++;
            }

            // Assign right child if value is not null
            if (i < values.length) {
                if (values[i] != null) {
                    current.right = new TreeNode(values[i]);
                    queue.add(current.right);
                }
                i++;
            }
        }

        return root;
    }

    private static String treeToString(TreeNode root) {
        if (root == null) {
            return "Empty tree";
        }

        StringBuilder result = new StringBuilder();
        // Use a queue for level-order traversal
        java.util.Queue<TreeNode> queue = new java.util.LinkedList<>();
        queue.add(root);

        int level = 0;

        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            result.append("Level ").append(level + 1).append(": ");

            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();

                if (node == null) {
                    result.append("null ");
                } else {
                    result.append(node.val).append(" ");

                    // Add children to queue (including nulls for visualization)
                    queue.add(node.left);
                    queue.add(node.right);
                }
            }

            result.append("\n");
            level++;

            // Check if the next level has only nulls
            boolean allNull = true;
            for (TreeNode node : queue) {
                if (node != null) {
                    allNull = false;
                    break;
                }
            }

            if (allNull) {
                break;
            }
        }

        return result.toString();
    }

    private static String truncateStr(String s, int maxLength) {
        if (s.length() <= maxLength) {
            return s;
        }
        return s.substring(0, maxLength - 3) + "...";
    }

    private static String truncateStr(String s) {
        return truncateStr(s, MAX_STRING_LENGTH);
    }

    private static boolean assertSolution(Integer[] inputList, int expected) {
        soln outer = new soln();
        Solution solution = outer.new Solution();

        TreeNode tree = createBinaryTree(inputList);

        int output = solution.minDepth(tree);

        boolean assertPass = output == expected;

        String result = assertPass ? "PASS" : "FAIL";
        String formattedResult = assertPass ? greenText(result) : redText(result);

        System.out.println(formattedResult + ": Output: " + output + ", Input: " + truncateStr(Arrays.toString(inputList)));

        if (!assertPass) {
            System.out.println("\n=== ACTUAL TREE ===");
            System.out.println(treeToString(tree));
        }

        System.out.println("==========================================");

        return assertPass;
    }

    public static void main(String[] args) {

        // Test cases for Minimum Depth of Binary Tree problem

        // Edge cases
        assertSolution(new Integer[] { null },
                /* expected */ 0);

        assertSolution(new Integer[] { 1 },
                /* expected */ 1);

        // Basic cases
        assertSolution(new Integer[] { 3, 9, 20 },
                /* expected */ 2);

        assertSolution(new Integer[] { 3, 9, 20, null, null, 15, 7 },
                /* expected */ 2);

        // Unbalanced trees
        assertSolution(new Integer[] { 2, null, 3, null, 4, null, 5, null, 6 },
                /* expected */ 5);

        // assertSolution(new Integer[] { 1, 2, null, 3, null, 4, null, 5 },
        //         /* expected */ 5);

        // // Trees with different depths on left and right
        // assertSolution(new Integer[] { 1, 2, 3, 4, null, null, 5, 6, null, null, 7 },
        //         /* expected */ 3);

        // assertSolution(new Integer[] { 1, 2, 3, 4, null, 5, 6, null, null, 7 },
        //         /* expected */ 2);

        // // Complete balanced tree
        // assertSolution(new Integer[] { 1, 2, 3, 4, 5, 6, 7 },
        //         /* expected */ 3);

        // // Complete tree with multiple levels
        // assertSolution(new Integer[] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15 },
        //         /* expected */ 4);

        // // Trees with negative values (values shouldn't affect the depth calculation)
        // assertSolution(new Integer[] { -10, -20, -30 },
        //         /* expected */ 2);

        // // Large tree (edge of constraints)
        // Integer[] largeTree = new Integer[100000];
        // largeTree[0] = 1;
        // for (int i = 1; i < 100000; i++) {
        //     if (i % 2 == 1) { // Left child
        //         largeTree[i] = i + 1;
        //     } else { // Right child is null
        //         largeTree[i] = null;
        //     }
        // }
        // assertSolution(largeTree,
        //         /* expected */ 50000);

        // // Tree with only left children
        // assertSolution(new Integer[] { 1, 2, null, 3, null, 4, null, 5, null },
        //         /* expected */ 5);

        // // Tree with only right children
        // assertSolution(new Integer[] { 1, null, 2, null, 3, null, 4, null, 5 },
        //         /* expected */ 5);

        // // Tree with multiple potential minimum paths
        // assertSolution(new Integer[] { 1, 2, 3, null, 4, null, 5, null, null, 6 },
        //         /* expected */ 3);

        // // Tree with leaf nodes at different levels
        // assertSolution(new Integer[] { 1, 2, 3, 4, 5, null, 6, null, null, 7, 8 },
        //         /* expected */ 3);

        // // Tree with values at the boundaries of constraints
        // assertSolution(new Integer[] { 1000, -1000, 0 },
        //         /* expected */ 2);

        // // More complex tree structures
        // assertSolution(new Integer[] { 5, 4, 8, 11, null, 13, 4, 7, 2, null, null, null, 1 },
        //         /* expected */ 3);

        // assertSolution(new Integer[] { 3, 1, 4, null, 2, null, 5 },
        //         /* expected */ 3);

    }
}
